
<img src="../../img/ods_stickers.jpg">

## <center> [mlcourse.ai](https://mlcourse.ai) – открытый курс OpenDataScience по машинному обучению 

Автор материала: Юрий Кашницкий (@yorko в Slack ODS). Материал распространяется на условиях лицензии [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/). Можно использовать в любых целях (редактировать, поправлять и брать за основу), кроме коммерческих, но с обязательным упоминанием автора материала.

# <center>Домашнее задание № 1 (демо).<br> Анализ данных по доходу населения UCI Adult</center>

**В задании предлагается с помощью Pandas ответить на несколько вопросов по данным репозитория UCI [Adult](https://archive.ics.uci.edu/ml/datasets/Adult) (качать данные не надо – они уже есть в репозитории). Для отправки решений используйте [онлайн-форму](https://docs.google.com/forms/d/1xAzU-5N6oEeR4UG8G44V6XL-Kbr4WYcPBG45Kijge2Y).**

Уникальные значения признаков (больше информации по ссылке выше):
- age: continuous.
- workclass: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.
- fnlwgt: continuous.
- education: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.
- education-num: continuous.
- marital-status: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.
- occupation: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.
- relationship: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.
- race: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.
- sex: Female, Male.
- capital-gain: continuous.
- capital-loss: continuous.
- hours-per-week: continuous.
- native-country: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.   
- salary: >50K,<=50K

In [2]:
import pandas as pd

In [3]:
data = pd.read_csv("adult.data.csv")
data.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

**1. Сколько мужчин и женщин (признак *sex*) представлено в этом наборе данных?**

In [4]:
data['sex']

0          Male
1          Male
2          Male
3          Male
4        Female
          ...  
32556    Female
32557      Male
32558    Female
32559      Male
32560    Female
Name: sex, Length: 32561, dtype: object

In [5]:
data[data['sex'] == 'Female']

age     workclass  fnlwgt     education  education-num  \
4       28       Private  338409     Bachelors             13   
5       37       Private  284582       Masters             14   
6       49       Private  160187           9th              5   
8       31       Private   45781       Masters             14   
12      23       Private  122272     Bachelors             13   
...    ...           ...     ...           ...            ...   
32546   37       Private  198216    Assoc-acdm             12   
32549   43     State-gov  255835  Some-college             10   
32556   27       Private  257302    Assoc-acdm             12   
32558   58       Private  151910       HS-grad              9   
32560   52  Self-emp-inc  287927       HS-grad              9   

              marital-status       occupation    relationship   race     sex  \
4         Married-civ-spouse   Prof-specialty            Wife  Black  Female   
5         Married-civ-spouse  Exec-managerial            Wife  White  Female   
6      Married-spouse-absent    Other-service   Not-in-family  Black  Female   
8              Never-married   Prof-specialty   Not-in-family  White  Female   
12             Never-married     Adm-clerical       Own-child  White  Female   
...                      ...              ...             ...    ...     ...   
32546               Divorced     Tech-support   Not-in-family  White  Female   
32549               Divorced     Adm-clerical  Other-relative  White  Female   
32556     Married-civ-spouse     Tech-support            Wife  White  Female   
32558                Widowed     Adm-clerical       Unmarried  White  Female   
32560     Married-civ-spouse  Exec-managerial            Wife  White  Female   

       capital-gain  capital-loss  hours-per-week native-country salary  
4                 0             0              40           Cuba  <=50K  
5                 0             0              40  United-States  <=50K  
6                 0             0              16        Jamaica  <=50K  
8             14084             0              50  United-States   >50K  
12                0             0              30  United-States  <=50K  
...             ...           ...             ...            ...    ...  
32546             0             0              40  United-States  <=50K  
32549             0             0              40  United-States  <=50K  
32556             0             0              38  United-States  <=50K  
32558             0             0              40  United-States  <=50K  
32560         15024             0              40  United-States   >50K  

[10771 rows x 15 columns]

### Ответ: 32561

In [6]:
data.describe()

age        fnlwgt  education-num  capital-gain  capital-loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours-per-week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

**2. Каков средний возраст (признак *age*) женщин?**

In [7]:
data[['age','sex']]

age     sex
0       39    Male
1       50    Male
2       38    Male
3       53    Male
4       28  Female
...    ...     ...
32556   27  Female
32557   40    Male
32558   58  Female
32559   22    Male
32560   52  Female

[32561 rows x 2 columns]

In [8]:
data[data['sex'] == 'Female']['age'].mean()

36.85823043357163

### Ответ: 36.85823043357163

**3. Какова доля граждан Германии (признак *native-country*)?**

In [9]:
data['native-country'].value_counts()

United-States                 29170
Mexico                          643
?                               583
Philippines                     198
Germany                         137
Canada                          121
Puerto-Rico                     114
El-Salvador                     106
India                           100
Cuba                             95
England                          90
Jamaica                          81
South                            80
China                            75
Italy                            73
Dominican-Republic               70
Vietnam                          67
Guatemala                        64
Japan                            62
Poland                           60
Columbia                         59
Taiwan                           51
Haiti                            44
Iran                             43
Portugal                         37
Nicaragua                        34
Peru                             31
France                      

In [10]:
137 / 32561

0.004207487485028101

### Ответ: 0.004207487485028101

**4-5. Каковы средние значения и среднеквадратичные отклонения возраста тех, кто получает более 50K в год (признак *salary*) и тех, кто получает менее 50K в год? **

In [11]:
data[data['salary'] == '<=50K']['age'].mean()

36.78373786407767

In [12]:
data[data['salary'] == '<=50K']['age'].std()

14.020088490824813

In [13]:
data['salary'].value_counts()

<=50K    24720
>50K      7841
Name: salary, dtype: int64

In [14]:
data[data['salary'] == '>50K']['age'].mean()

44.24984058155847

In [15]:
data[data['salary'] == '>50K']['age'].std()

10.51902771985177

### Ответ: <= 50K средний возраст - 36.78373786407767 sd - 14.020088490824813
### >50K средний возраст -  44.24984058155847 sd - 10.51902771985177

**6. Правда ли, что люди, которые получают больше 50k, имеют как минимум высшее образование? (признак *education – Bachelors, Prof-school, Assoc-acdm, Assoc-voc, Masters* или *Doctorate*)**

In [16]:
data[data['salary'] == '>50K']['education'].value_counts()

Bachelors       2221
HS-grad         1675
Some-college    1387
Masters          959
Prof-school      423
Assoc-voc        361
Doctorate        306
Assoc-acdm       265
10th              62
11th              60
7th-8th           40
12th              33
9th               27
5th-6th           16
1st-4th            6
Name: education, dtype: int64

### Нет, не правда

**7. Выведите статистику возраста для каждой расы (признак *race*) и каждого пола. Используйте *groupby* и *describe*. Найдите таким образом максимальный возраст мужчин расы *Amer-Indian-Eskimo*.**

In [17]:
data.groupby('race')['age'].describe()

count       mean        std   min   25%   50%   75%  \
race                                                                        
Amer-Indian-Eskimo    311.0  37.173633  12.447130  17.0  28.0  35.0  45.5   
Asian-Pac-Islander   1039.0  37.746872  12.825133  17.0  28.0  36.0  45.0   
Black                3124.0  37.767926  12.759290  17.0  28.0  36.0  46.0   
Other                 271.0  33.457565  11.538865  17.0  25.0  31.0  41.0   
White               27816.0  38.769881  13.782306  17.0  28.0  37.0  48.0   

                     max  
race                      
Amer-Indian-Eskimo  82.0  
Asian-Pac-Islander  90.0  
Black               90.0  
Other               77.0  
White               90.0

### Ответ: 82

**8. Среди кого больше доля зарабатывающих много (>50K): среди женатых или холостых мужчин (признак *marital-status*)? Женатыми считаем тех, у кого *marital-status* начинается с *Married* (Married-civ-spouse, Married-spouse-absent или Married-AF-spouse), остальных считаем холостыми.**

In [18]:
data[data['salary'] == '>50K']['marital-status'].value_counts()

Married-civ-spouse       6692
Never-married             491
Divorced                  463
Widowed                    85
Separated                  66
Married-spouse-absent      34
Married-AF-spouse          10
Name: marital-status, dtype: int64

In [19]:
data[data['salary'] == '>50K'][data['marital-status'] == 'Married-civ-spouse']

C:\Users\Alexandr\AppData\Local\Temp\ipykernel_8244\868343628.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  data[data['salary'] == '>50K'][data['marital-status'] == 'Married-civ-spouse']


age         workclass  fnlwgt     education  education-num  \
7       52  Self-emp-not-inc  209642       HS-grad              9   
9       42           Private  159449     Bachelors             13   
10      37           Private  280464  Some-college             10   
11      30         State-gov  141297     Bachelors             13   
14      40           Private  121772     Assoc-voc             11   
...    ...               ...     ...           ...            ...   
32539   71                 ?  287372     Doctorate             16   
32545   39         Local-gov  111499    Assoc-acdm             12   
32554   53           Private  321865       Masters             14   
32557   40           Private  154374       HS-grad              9   
32560   52      Self-emp-inc  287927       HS-grad              9   

           marital-status         occupation relationship                race  \
7      Married-civ-spouse    Exec-managerial      Husband               White   
9      Married-civ-spouse    Exec-managerial      Husband               White   
10     Married-civ-spouse    Exec-managerial      Husband               Black   
11     Married-civ-spouse     Prof-specialty      Husband  Asian-Pac-Islander   
14     Married-civ-spouse       Craft-repair      Husband  Asian-Pac-Islander   
...                   ...                ...          ...                 ...   
32539  Married-civ-spouse                  ?      Husband               White   
32545  Married-civ-spouse       Adm-clerical         Wife               White   
32554  Married-civ-spouse    Exec-managerial      Husband               White   
32557  Married-civ-spouse  Machine-op-inspct      Husband               White   
32560  Married-civ-spouse    Exec-managerial         Wife               White   

          sex  capital-gain  capital-loss  hours-per-week native-country  \
7        Male             0             0              45  United-States   
9        Male          5178             0              40  United-States   
10       Male             0             0              80  United-States   
11       Male             0             0              40          India   
14       Male             0             0              40              ?   
...       ...           ...           ...             ...            ...   
32539    Male             0             0              10  United-States   
32545  Female             0             0              20  United-States   
32554    Male             0             0              40  United-States   
32557    Male             0             0              40  United-States   
32560  Female         15024             0              40  United-States   

      salary  
7       >50K  
9       >50K  
10      >50K  
11      >50K  
14      >50K  
...      ...  
32539   >50K  
32545   >50K  
32554   >50K  
32557   >50K  
32560   >50K  

[6692 rows x 15 columns]

### Ответ: среди женатых

**9. Какое максимальное число часов человек работает в неделю (признак *hours-per-week*)? Сколько людей работают такое количество часов и каков среди них процент зарабатывающих много?**

In [20]:
data.describe()

age        fnlwgt  education-num  capital-gain  capital-loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours-per-week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

In [21]:
data[data['hours-per-week'] == 99]

age         workclass  fnlwgt     education  education-num  \
935     37           Private  176900       HS-grad              9   
1172    25           Private  404616       Masters             14   
1887    55  Self-emp-not-inc  184425  Some-college             10   
3578    37      Self-emp-inc  382802     Doctorate             16   
4086    50                 ?  174964          10th              6   
...    ...               ...     ...           ...            ...   
30992   40      Self-emp-inc  104235       Masters             14   
31681   43           Private  266324       HS-grad              9   
31699   43  Self-emp-not-inc  163108  Some-college             10   
32469   58      Self-emp-inc  181974     Doctorate             16   
32531   30                 ?   33811     Bachelors             13   

           marital-status       occupation   relationship                race  \
935    Married-civ-spouse     Craft-repair        Husband               White   
1172   Married-civ-spouse  Farming-fishing  Not-in-family               White   
1887   Married-civ-spouse  Farming-fishing        Husband               White   
3578   Married-civ-spouse   Prof-specialty        Husband               Black   
4086   Married-civ-spouse                ?        Husband               White   
...                   ...              ...            ...                 ...   
30992       Never-married    Other-service      Own-child               White   
31681  Married-civ-spouse     Craft-repair        Husband               White   
31699  Married-civ-spouse  Farming-fishing        Husband               White   
32469       Never-married   Prof-specialty  Not-in-family               White   
32531       Never-married                ?  Not-in-family  Asian-Pac-Islander   

          sex  capital-gain  capital-loss  hours-per-week native-country  \
935      Male             0             0              99  United-States   
1172     Male             0             0              99  United-States   
1887     Male             0             0              99  United-States   
3578     Male             0             0              99  United-States   
4086     Male             0             0              99  United-States   
...       ...           ...           ...             ...            ...   
30992    Male             0             0              99  United-States   
31681    Male             0          1902              99  United-States   
31699    Male             0             0              99  United-States   
32469  Female             0             0              99              ?   
32531  Female             0             0              99  United-States   

      salary  
935     >50K  
1172    >50K  
1887    >50K  
3578    >50K  
4086   <=50K  
...      ...  
30992  <=50K  
31681   >50K  
31699  <=50K  
32469  <=50K  
32531  <=50K  

[85 rows x 15 columns]

In [22]:
data[data['hours-per-week'] == 99][data['salary'] == '>50K'].describe()

C:\Users\Alexandr\AppData\Local\Temp\ipykernel_8244\1715917269.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  data[data['hours-per-week'] == 99][data['salary'] == '>50K'].describe()


age         fnlwgt  education-num  capital-gain  capital-loss  \
count  25.000000      25.000000      25.000000     25.000000     25.000000   
mean   40.760000  188218.680000      11.680000   2235.040000    427.440000   
std     9.355391  115655.645292       3.009983   5052.657523    878.718009   
min    25.000000   24790.000000       4.000000      0.000000      0.000000   
25%    37.000000  111499.000000      10.000000      0.000000      0.000000   
50%    39.000000  178370.000000      11.000000      0.000000      0.000000   
75%    43.000000  266324.000000      14.000000      0.000000      0.000000   
max    73.000000  404616.000000      16.000000  15024.000000   2415.000000   

       hours-per-week  
count            25.0  
mean             99.0  
std               0.0  
min              99.0  
25%              99.0  
50%              99.0  
75%              99.0  
max              99.0

In [23]:
25 / 85

0.29411764705882354

### Ответ: 29.411764705882354%

**10. Посчитайте среднее время работы (*hours-per-week*) зарабатывающих мало и много (*salary*) для каждой страны (*native-country*).**

In [24]:
for country, sub_df in data.groupby('native-country'):
    print("Страна: " + country)
    print("Среднее время работы зарабатывающих <=50K: " + str(sub_df[sub_df['salary'] == '<=50K']['hours-per-week'].mean()))
    print("Среднее время работы зарабатывающих >50K: " + str(sub_df[sub_df['salary'] == '>50K']['hours-per-week'].mean()))
    print('-----------------------------------\n')

Страна: ?
Среднее время работы зарабатывающих <=50K: 40.16475972540046
Среднее время работы зарабатывающих >50K: 45.54794520547945
-----------------------------------

Страна: Cambodia
Среднее время работы зарабатывающих <=50K: 41.416666666666664
Среднее время работы зарабатывающих >50K: 40.0
-----------------------------------

Страна: Canada
Среднее время работы зарабатывающих <=50K: 37.91463414634146
Среднее время работы зарабатывающих >50K: 45.64102564102564
-----------------------------------

Страна: China
Среднее время работы зарабатывающих <=50K: 37.38181818181818
Среднее время работы зарабатывающих >50K: 38.9
-----------------------------------

Страна: Columbia
Среднее время работы зарабатывающих <=50K: 38.68421052631579
Среднее время работы зарабатывающих >50K: 50.0
-----------------------------------

Страна: Cuba
Среднее время работы зарабатывающих <=50K: 37.98571428571429
Среднее время работы зарабатывающих >50K: 42.44
-----------------------------------

Страна: Dominica